src/main/java/org/aksw/simba/lemming/creation/Inferer.java

```diff
@@ -0,0 +1,178 @@
+			set.add(curStat.getSubject());
+			if(curStat.getObject().isResource()) {
+				set.add(curStat.getObject().asResource());
+			} 
+		}
+		checkEmptyTypes(set, model);
+		return set;
+	}
+	
+	/**
+	 * This method simply logs the count of how many resources without a type exist in a given 
+	 * model
+	 * @param set group of resources that we want to check in the model
+	 * if a type relation is existing or not
+	 * @param model RDF Model where this needs to be checked in
+	 */
+	private void checkEmptyTypes(Set<Resource> set, Model model) {
+		int emptyTypeCount = 0;
+		for (Resource resource : set) {
+			if (!model.contains(resource, RDF.type)) {
+				emptyTypeCount++;
+			} 
+		}
+		LOGGER.info("Number of resources without type : " + emptyTypeCount);
+	}
+	
+	/**
+	 * This method iterates through the model's statements, continuously searching for each 
+	 * property in the ontology and adding the inferred triples to the new model
+	 * @param newModel model where we will add the new triples
+	 * @param sourceModel provided model where we iterate through the statements
+	 * @param ontModel the ontology model
+	 */
+	public void iterateStmts(Model newModel, Model sourceModel, OntModel ontModel) {
+		StmtIterator stmts = sourceModel.listStatements();
+		while(stmts.hasNext()) {
+			Statement curStatement = stmts.next();
+			List<Statement> stmtsList = new ArrayList<>();
+			Set<Statement> newStmts = searchType(curStatement, ontModel, newModel);
+			stmtsList.addAll(newStmts);
+			newModel.add(stmtsList);
+		}
+	}
+
+	/**
+	 * For a given statement, this method searches for the predicate of a model inside 
+	 * the Ontology. If found in the Ontology, it then extracts the domain and range.
+	 * Creating and adding a new triple with the inferred type to the model.
+	 * @param statement statement in which we want to check the predicate in the ontology
+	 * @param ontModel the ontology model
+	 * @param newModel where we add the new triples and therefore, where we check if the 
+	 * statement is already existing in the model or not
+	 * @return a set of statements inferred from one property
+	 */
+	private Set<Statement> searchType(Statement statement, OntModel ontModel, Model newModel) {
+		Set<Statement> newStmts = new HashSet<>();
+		Resource subject = statement.getSubject();
+		Property predicate = statement.getPredicate();
+		RDFNode object = statement.getObject();
+
+		// search for the predicate of the model in the ontology
+		OntProperty temp = ontModel.getOntProperty(predicate.toString());
+		if (temp != null) {
+			List<? extends OntResource> domain = temp.listDomain().toList();
+			for(OntResource curResource: domain){
+				Statement subjType = ResourceFactory.createStatement(subject, RDF.type, curResource);
+				if (!newModel.contains(subjType)) {
+					newStmts.add(subjType);
+				}
+			}
+			
+			List<? extends OntResource> range = temp.listRange().toList();
+			for(OntResource curResource: range){
+				if(object.isResource()) {

```

Comment:
It would make sense to move this `if` statement a little bit more "up". I think it should be already checked before asking for the range of the property. If the object is not a resource, we are not interested in the range, anyway :wink: